bolt/lib/Passes/ProfileQualityStats.cpp

```diff
@@ -0,0 +1,575 @@
+    std::sort(NumUnreachables.begin(), NumUnreachables.end());
+    printDistribution(OS, NumUnreachables);
+
+    OS << "distribution of SUM_EC(unreachable POS BBs) per function\n";
+    std::sort(SumECUnreachables.begin(), SumECUnreachables.end());
+    printDistribution(OS, SumECUnreachables);
+
+    OS << "distribution of [(SUM_EC(unreachable POS BBs) / SUM_EC(all "
+          "POS BBs))] per function\n";
+    printDistribution(OS, FractionECUnreachables, /*Fraction=*/true);
+  }
+}
+
+void printCallGraphFlowConservationStats(
+    raw_ostream &OS, iterator_range<function_iterator> &Functions,
+    FlowInfo &TotalFlowMap) {
+  std::vector<double> CallGraphGaps;
+
+  for (const BinaryFunction *Function : Functions) {
+    if (Function->size() <= 1 || !Function->isSimple())
+      continue;
+
+    const uint64_t FunctionNum = Function->getFunctionNumber();
+    FlowMapTy &IncomingMap = TotalFlowMap.TotalIncomingMaps[FunctionNum];
+    FlowMapTy &OutgoingMap = TotalFlowMap.TotalOutgoingMaps[FunctionNum];
+    FunctionFlowMapTy &CallGraphIncomingMap = TotalFlowMap.CallGraphIncomingMap;
+
+    // Only consider functions that are not a program entry.
+    if (CallGraphIncomingMap.find(FunctionNum) != CallGraphIncomingMap.end()) {
+      uint64_t EntryInflow = 0;
+      uint64_t EntryOutflow = 0;
+      uint32_t NumConsideredEntryBlocks = 0;
+      for (const BinaryBasicBlock &BB : *Function) {
+        if (BB.isEntryPoint()) {
+          // If entry is an exit, then we don't consider it for flow
+          // conservation
+          if (BB.succ_size() == 0)
+            continue;
+          NumConsideredEntryBlocks++;
+
+          EntryInflow += IncomingMap[BB.getLayoutIndex()];
+          EntryOutflow += OutgoingMap[BB.getLayoutIndex()];
+        }
+      }
+      uint64_t NetEntryOutflow = 0;
+      if (EntryOutflow < EntryInflow) {
+        if (opts::Verbosity >= 2) {
+          // We expect entry blocks' CFG outflow >= inflow, i.e., it has a
+          // non-negative net outflow. If this is not the case, then raise a
+          // warning if requested.
+          OS << "BOLT WARNING: unexpected entry block CFG outflow < inflow "
+                "in "
+                "function "
+             << Function->getPrintName() << "\n";
+          if (opts::Verbosity >= 3)
+            Function->dump();
+        }
+      } else {
+        NetEntryOutflow = EntryOutflow - EntryInflow;
+      }
+      if (NumConsideredEntryBlocks > 0) {
+        const uint64_t CallGraphInflow =
+            TotalFlowMap.CallGraphIncomingMap[Function->getFunctionNumber()];
+        const uint64_t Min = std::min(NetEntryOutflow, CallGraphInflow);
+        const uint64_t Max = std::max(NetEntryOutflow, CallGraphInflow);
+        const double CallGraphGap = 1 - (double)Min / Max;
+
+        if (opts::Verbosity >= 2 && CallGraphGap >= 0.5) {
+          OS << "Nontrivial call graph gap of size "
+             << format("%.2lf%%", 100 * CallGraphGap)
```

Comment:
Here and in other places
> For example, to format a floating point value as a percentage, you can use the style option P.

https://llvm.org/docs/ProgrammersManual.html#simple-formatting